src/recordlinker/linking/mpi_service.py

```diff
@@ -53,7 +53,22 @@ def get_block_data(
         )
 
     # Using the subquery of unique Patient IDs, select all the Patients
-    expr = expression.select(models.Patient).where(models.Patient.id.in_(base))
+    candidate_patients = expression.select(models.Patient).where(models.Patient.id.in_(base))
+    return candidate_patients
```

Comment:
@ericbuckley Alternatively, could we keep `get_block_data()` signatures as usual and change to the below with [`orm.lazyload()`](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.lazyload)?
```suggestion
    expr = expression.select(models.Patient).where(models.Patient.id.in_(base))
    return orm.lazyload(session.execute(expr).scalars().all())
```

Reply:
Lazy loading isn't quite the same as lazy evaluation.  The former is about retrieving foreign key relationships, so for example if you get back a Patient object and call object.person.  The person query is lazy loaded and triggers another SQL query to retrieve the person info from the database.

Lazy evaluation is about not executing the SQL statement until you attempt to iterate over the results.  `session.execute(expr).scalars()` returns a sql result.  You can pass that around and not have the SQL statement execute, however as soon as you do something like `len([x for x in result])`, the SQL would be executed before the length was calculated.  However `session.execute(expr).scalars().all()` executes immediately, and passes a sequence object around with the results already stored in memory.  

This however doesn't work as I hoped it might in sqlalchemy.  As even if you return a SQL result, and pass that  into a new function to execute a query, its going to execute that result before executing the second query, thus resulting in two queries anyways.  Let me try to explain in a simple example

```python
result = get_persons(id=[1,2,3])
return get_patients(persons=result)
```

Take the above code, a function that returns a collection of persons that have ids 1, 2 or 3.  Then another function that returns all the patients that are within those person clusters.

Ideally the ORM would know it can create SQL for the second function like so, which would take the lazy result and create a subquery out of it, thus resulting in 1 total query
```sql
SELECT pat.*
FROM patients pat
WHERE pat.person_id IN (SELECT * FROM persons WHERE id in [1,23])
```

However, sqlalchemy isn't going to do that.  Instead of converting the result into a subquery, its going to evaluate it first, then pass it in, thus resulting in 2 executed queries.
```sql
SELECT pat.*
FROM patients pat
WHERE pat.person_id IN ([1,23])
```

The above example is contrived, and we of course don't need to call two functions if we already have a list of person ids, but hopefully it shows the differences in the SQL queries generated.

Reply:
Gotcha, I think this makes sense. In other words, the behavior we had hoped for by passing around the sql result `session.execute(expr).scalars()` without it executing is not the case with SQLAlchemy (resulting in 2 queries anyways).

In that case, does it make more sense now to take the alternative approach in just modifying the one query in `get_block_data()`?